### ISSUE TITLE
Unhandled error during execution of scheduler flush related to NuxtPage, useAsyncData & refs 

### ISSUE THREAD

**Issue by Striffly** [nuxt/nuxt#21901]
### Environment

------------------------------
- Operating System: Linux
- Node Version:     v18.12.1
- Nuxt Version:     3.6.1
- Nitro Version:    2.5.2
- Package Manager:  yarn@1.22.19
- Builder:          vite
- User Config:      devtools
- Runtime Modules:  -
- Build Modules:    -
------------------------------

### Reproduction

Reproduction : https://stackblitz.com/edit/github-88hbfa?file=pages%2Findex.vue

Steps to reproduce:
1. In a page file, retrieve data using `useAsyncData` and display an array of elements in the template using this data.
2. Assign the array of elements to a ref variable as described in the Vue.js documentation: https://vuejs.org/guide/essentials/template-refs.html#accessing-the-refs.
3. Display this page in a browser.
4. Edit the JS code of this page (e.g., modify console.log).
5. Errors will appear.

This error does not occur if:
- The page is being shown for the first time (no reloading through JS edits).
- The code specified above is inside app.vue and not inside a page that is loaded with NuxtPage (if you copy/paste index.vue inside app.vue, errors do not occur).
- No ref is used
- `useAsyncData` is not used

### Describe the bug

Console error is shown related to scheduler flush.
Also, `TypeError: Cannot read properties of null (reading 'parentNode')` error is thrown.
See Logs field for more detials about these errors

### Workarounds

- Reload page (meh)
- Use `reactive` instead of `ref`
```html
<template>
<div>
  <li v-for="menuItem in menuItems" ref="(el) => updateMenuItemRefs(el)"></li>
</div>
</template>

<script setup>
[...]
const menuItemRefs = reactive([])
function updateMenuItemRefs (el) {
  menuItemRefs.push(el)
}
</script>
```

### Logs

(initially made with `test.vue` page, but error is the same)

```shell-script
test.vue:14 edit this message 2

chunk-3Q27KRJ5.js:97 [Vue warn]: Unhandled error during execution of scheduler flush. This is likely a Vue internals bug. Please open an issue at https://new-issue.vuejs.org/?repo=vuejs/core 
  at <RouteProvider key="/test" routeProps= {Component: {…}, route: {…}} pageKey="/test"  ... > 
  at <RouterView name=undefined route=undefined > 
  at <NuxtPage> 
  at <NuxtLayout> 
  at <App key=3 > 
  at <NuxtRoot>
warn @ chunk-3Q27KRJ5.js:97
logError @ chunk-3Q27KRJ5.js:271
handleError @ chunk-3Q27KRJ5.js:263
callWithErrorHandling @ chunk-3Q27KRJ5.js:215
flushJobs @ chunk-3Q27KRJ5.js:411
Promise.then (async)
queueFlush @ chunk-3Q27KRJ5.js:324
queueJob @ chunk-3Q27KRJ5.js:318
reload @ chunk-3Q27KRJ5.js:517
(anonymous) @ chunk-3Q27KRJ5.js:547
(anonymous) @ test.vue?t=1688346524889:81
(anonymous) @ client.ts:542
(anonymous) @ client.ts:459
(anonymous) @ client.ts:306
queueUpdate @ client.ts:306
await in queueUpdate (async)
(anonymous) @ client.ts:159
handleMessage @ client.ts:157
(anonymous) @ client.ts:90
Show 7 more frames

chunk-MNWBZQKJ.js:82 Uncaught (in promise) TypeError: Cannot read properties of null (reading 'parentNode')
    at parentNode (chunk-MNWBZQKJ.js:82:30)
    at ReactiveEffect.componentUpdateFn [as fn] (chunk-3Q27KRJ5.js:5823:11)
    at ReactiveEffect.run (reactivity.esm-bundler.js:178:19)
    at instance.update (chunk-3Q27KRJ5.js:5860:52)
    at callWithErrorHandling (chunk-3Q27KRJ5.js:213:32)
    at flushJobs (chunk-3Q27KRJ5.js:411:9)
parentNode @ chunk-MNWBZQKJ.js:82
componentUpdateFn @ chunk-3Q27KRJ5.js:5823
run @ reactivity.esm-bundler.js:178
instance.update @ chunk-3Q27KRJ5.js:5860
callWithErrorHandling @ chunk-3Q27KRJ5.js:213
flushJobs @ chunk-3Q27KRJ5.js:411
Promise.then (async)
queueFlush @ chunk-3Q27KRJ5.js:324
queueJob @ chunk-3Q27KRJ5.js:318
reload @ chunk-3Q27KRJ5.js:517
(anonymous) @ chunk-3Q27KRJ5.js:547
(anonymous) @ test.vue?t=1688346524889:81
(anonymous) @ client.ts:542
(anonymous) @ client.ts:459
(anonymous) @ client.ts:306
queueUpdate @ client.ts:306
await in queueUpdate (async)
(anonymous) @ client.ts:159
handleMessage @ client.ts:157
(anonymous) @ client.ts:90
Show 8 more frames
```

**Comment by boboldehampsink** [comment]
Seeing the same issue

**Comment by Striffly** [comment]
> Duplicate of #21841?

I don't think so, because the context differs in many ways compared to this issue.

**Comment by chris-lsn** [comment]
Same on v3.6.5, it works if useAsyncData isn't used or on first time visit or after refreshing the page
Temporary rollback to v3.6.0 to bypass this issue.

**Comment by danielroe** [comment]
Is there a chance you can reproduce with https://stackblitz.com/github/nuxt-contrib/vue3-ssr-starter ?

**Comment by wiseadme** [comment]
Faced the same problem in version 3.8.0

**Comment by number304** [comment]
Same problem in 3.8.2

**Comment by mrc-bsllt** [comment]
Same problem in version 3.9.1
Mine is caused by the layout change, and the layout is set in the pages via definePageMeta(). Does anyone know how to solve it?

**Comment by dm-heinze** [comment]
Got the same problem in 3.9.3 also caused by definePageMeta(), but used to define a route middleware that redirects to another route.

**Comment by WJakub** [comment]
> I was having a similar issue. In my project I am using a custom layout and solved it by using `<Suspense>` on the page itself (wrapping my custom layout): `<div><Suspense><NuxtLayout name="two-columns">...</NuxtLayout></Suspense></div>`. So the whole page is waiting on each await in any component. Maybe this helps someone else as well.

Thank you. In my case, wrapping `NuxtLayout` in the `Suspense` component within my entry `app.vue` was not enough to fix the flush problems.

However, after wrapping `<slot />` with the `Suspense` component within my layout components, the problem was resolved.

**Comment by EntrenaEnCasa** [comment]
I am facing the same issue. It seems to be related with layouts. Still no fixes for this issue? is it Vue related and not Nuxt related?

**Comment by professorhaseeb** [comment]
> > I was having a similar issue. In my project I am using a custom layout and solved it by using `<Suspense>` on the page itself (wrapping my custom layout): `<div><Suspense><NuxtLayout name="two-columns">...</NuxtLayout></Suspense></div>`. So the whole page is waiting on each await in any component. Maybe this helps someone else as well.
> 
> Thank you. In my case, wrapping `NuxtLayout` in the `Suspense` component within my entry `app.vue` was not enough to fix the flush problems.
> 
> However, after wrapping `<slot />` with the `Suspense` component within my layout components, the problem was resolved.

```
<Suspense>
    <slot />
</Suspense>
```

this solves this issue

**Comment by luca-smartpricing** [comment]
Is there any downgrade adding the `<Suspense>` around `<slot />`? 
It affects the whole site and I don't want any surprises

**Comment by Gonzalo-Bruna** [comment]
So the reproduction steps that I found in my app are as follows:

- Go into any page which has a API call
- log out from my app (this changes layouts)
- log in back again (this changes layouts again) 

and the pages which also have api calls throw both this errors:

`Unhandled error during execution of render function`

`Unhandled error during execution of scheduler flush. This is likely a Vue internals bug. Please open an issue at https://github.com/vuejs/core .`

by changing the first step (not going to a page which has an API call) the issue never triggers, so I wonder if it might be related to the way of handling async data

**Comment by blqke** [comment]
If it helps, I was having this error when I was switching route (and layout) too fast (in my case there are 2 slots on my layout, left column and right column, and user is able to select item from left col with keyboard cf [this video](https://www.youtube.com/watch?v=aT_A6KSYACM)), setting the shortcuts inside page `onMounted` was the solution.

**Comment by luca-smartpricing** [comment]
It was fixed in Vue 3.4.20 [#10098](https://github.com/vuejs/core/issues/10098#issue-2079176480)

**Comment by frankan0** [comment]
> > I was having a similar issue. In my project I am using a custom layout and solved it by using `<Suspense>` on the page itself (wrapping my custom layout): `<div><Suspense><NuxtLayout name="two-columns">...</NuxtLayout></Suspense></div>`. So the whole page is waiting on each await in any component. Maybe this helps someone else as well.
> 
> Thank you. In my case, wrapping `NuxtLayout` in the `Suspense` component within my entry `app.vue` was not enough to fix the flush problems.
> 
> However, after wrapping `<slot />` with the `Suspense` component within my layout components, the problem was resolved.

it works  for me,thanks .

**Comment by rutbergphilip** [comment]
> It was fixed in Vue 3.4.20 [#10098](https://github.com/vuejs/core/issues/10098#issue-2079176480)

I have this same issue once again, using:
Nuxt `3.12.4`
Vue `3.4.33`

Downgrading to Vue `3.4.31` solved the issue for now.

**Comment by danielroe** [comment]
In that case, would you see if you can reproduce this in pure Vue SSR without Nuxt ([reproduction sandbox](https://stackblitz.com/github/nuxt-contrib/vue3-ssr-starter)), and if so, raise [there](https://github.com/vuejs/core/issues/new/choose)? 🙏

**Comment by rutbergphilip** [comment]
> and

I wasn't able to replicate this in the Vue-only environment, although I will admit I not used to that reproduction environment..

I was however able to trace the issue away from any relation to useAsyncData and into the `error.vue` file.

Adding a `<UNotifications />` into the error.vue file in latest version of Vue causes the following error to be thrown `TypeError: can't access property "parentNode", child is null`.
```html
<template>
  <div>
    <UMain>
      <UContainer>
        <UPage>
          <UPageError :error="error" />
        </UPage>
      </UContainer>
    </UMain>

    <UNotifications />
  </div>
</template>
```

By simply removing the `<UNotifications />` from this file resolves this issue, so I suppose this could actually be nuxt/ui related issue. I'll try to setup a reproduction sandbox for this.
